### ISSUE TITLE
Hysteria2 inbound/server memory leak 1.9.3 - 1.10.0-alpha.29

### ISSUE THREAD

**Issue by SasukeFreestyle** [SagerNet/sing-box#2027]
### Operating system

Linux

### System version

Ubuntu 22.04

### Installation type

Original sing-box Command Line

### Version

```shell
sing-box version 1.10.0-alpha.29

Environment: go1.22.6 linux/amd64
Tags: with_gvisor,with_quic,with_dhcp,with_wireguard,with_ech,with_utls,with_reality_server,with_acme,with_clash_api
CGO: disabled
```

### Description

I've been getting memory leaks from version 1.9.3 (maybe even earlier) and above.
sing-box exceeds well over 2GB of RAM after ~8 hrs of runtime, after some time it crashes. I've currently no log of the crash but can update issue when I've a log from this.

I'm still learning go so bear with me :)

btop:
![image](https://github.com/user-attachments/assets/ea8937d0-1fbd-42c9-8207-d57abea1e1ad)

goroutine:
![goroutine](https://github.com/user-attachments/assets/34efa9de-4007-4b83-9b08-920af1ece4a3)

heap:
![heap](https://github.com/user-attachments/assets/dcc49880-66d5-41f2-a5e8-91b2f603b151)

pprof dumps:
[pprof.heap.pb.gz](https://github.com/user-attachments/files/16624898/pprof.heap.pb.gz)

[pprof.goroutine.pb.gz](https://github.com/user-attachments/files/16624899/pprof.goroutine.pb.gz)

### Reproduction

This has occurred on all machines I've tested hysteria2 inbound on so it should be easy to reproduce. I would guess you need about ~100 users and after about ~12 hours of uptime

Server certs are generated by openssl.

Server config:
```json
{
   "log":{
      "level":"fatal",
      "timestamp":true
   },
   "experimental":{
      "cache_file":{
         "enabled":true
      }
   },
   "inbounds":[
      {
         "type":"hysteria2",
         "tag":"hy2-in-443",
         "listen":"127.0.0.1",
         "udp_timeout":180,
         "listen_port":443,
         "sniff":true,
         "sniff_override_destination":true,
         "domain_strategy":"prefer_ipv4",
         "up_mbps":0,
         "down_mbps":0,
         "obfs":{
            "type":"salamander",
            "password":"123"
         },
         "users":[
            {
               "name":"user",
               "password":"123"
            }
         ],
         "ignore_client_bandwidth":true,
         "tls":{
            "enabled":true,
            "certificate_path":"ca.crt",
            "key_path":"ca.key"
         }
      }
   ],
   "outbounds":[
      {
         "type":"direct",
         "tag":"direct"
      },
      {
         "type":"block",
         "tag":"block"
      },
      {
         "type":"dns",
         "tag":"dns-out"
      }
   ],
   "dns":{
      "disable_cache":true,
      "servers":[
         {
            "tag":"dns-out",
            "address":"udp://127.0.0.53",
            "address_strategy":"prefer_ipv4",
            "strategy":"prefer_ipv4",
            "detour":"direct"
         }
      ]
   },
   "route":{
      "rules":[
         {
            "ip_is_private":true,
            "outbound":"block"
         },
         {
            "rule_set":[
               "geoip-cn",
               "geoip-ir",
               "geoip-ru",
               "geoip-phishing",
               "geoip-malware",
               "geoip-private",
               "geosite-ir",
               "geosite-malware",
               "geosite-cryptominers",
               "geosite-phishing"
            ],
            "outbound":"block"
         },
         {
            "protocol":"dns",
            "outbound":"dns-out"
         }
      ],
      "rule_set":[
         {
            "tag":"geoip-cn",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geoip-cn.srs"
         },
         {
            "tag":"geoip-ir",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geoip-ir.srs"
         },
         {
            "tag":"geoip-ru",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geoip-ru.srs"
         },
         {
            "tag":"geoip-private",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geoip-private.srs"
         },
         {
            "tag":"geosite-ir",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://github.com/bootmortis/sing-geosite/releases/latest/download/geosite-ir.srs"
         },
         {
            "tag":"geoip-phishing",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geoip-phishing.srs"
         },
         {
            "tag":"geosite-phishing",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geosite-phishing.srs"
         },
         {
            "tag":"geoip-malware",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geoip-malware.srs"
         },
         {
            "tag":"geosite-malware",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geosite-malware.srs"
         },
         {
            "tag":"geosite-cryptominers",
            "type":"remote",
            "format":"binary",
            "update_interval":"3d",
            "url":"https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/geosite-cryptominers.srs"
         }
      ],
      "auto_detect_interface":true
   }
}
```

### Supporter

- [ ] I am a [sponsor](https://github.com/sponsors/nekohasekai/)

### Integrity requirements

- [X] I confirm that I have read the documentation, understand the meaning of all the configuration items I wrote, and did not pile up seemingly useful options or default values.
- [x] I confirm that I have provided the server and client configuration files and process that can be reproduced locally, instead of a complicated client configuration file that has been stripped of sensitive data.
- [X] I confirm that I have provided the simplest configuration that can be used to reproduce the error I reported, instead of depending on remote servers, TUN, graphical interface clients, or other closed-source software.
- [X] I confirm that I have provided the complete configuration files and logs, rather than just providing parts I think are useful out of confidence in my own intelligence.

**Comment by nekohasekai** [comment]
I guess the problem was introduced in sing-box 1.10.0-alpha.23. Can you check if the problem exists in versions before alpha 23 or 1.9.3?

**Comment by SasukeFreestyle** [comment]
I'm sure the problem occurred in 1.9.3 and also versions below see https://github.com/SagerNet/sing-box/issues/1245
With versions 1.9.0+ connections are properly closed, but there is still a leak.
Im running 1.9.3 now but it will take some hours for the leak to reproduce again. In the next post I will dump pprof of 1.9.3

**Comment by SasukeFreestyle** [comment]
So here are the 1.9.3 dumps, took longer than expected. Uptime 1 day 14 hours.

heap:
![heap1](https://github.com/user-attachments/assets/e0e8b8de-be74-4ca8-95c0-813d663726e6)

goroutine:
![goroutine1](https://github.com/user-attachments/assets/3bc7f982-e4a6-4963-823b-158387fc0a59)


heap:
[pprof.heap.pb.gz](https://github.com/user-attachments/files/16645521/pprof.heap.pb.gz)
goroutine:
[pprof.goroutine.pb.gz](https://github.com/user-attachments/files/16645524/pprof.goroutine.pb.gz)

So I've switched to alpha 22 as of writing this and will dump those in the next post.

Thank you for your time and effort.

**Comment by SasukeFreestyle** [comment]
1.10.0-alpha.22

![image](https://github.com/user-attachments/assets/01111b8b-f372-4b5d-9477-e7a9f08be402)

heap:
![heap22](https://github.com/user-attachments/assets/0b211d19-59af-4acd-a32d-4ce28e63c134)

goroutine:

![goroutine22](https://github.com/user-attachments/assets/593a3da9-8561-4c99-94bb-d19760d5ed85)

heap:
[pprof.heap.pb.gz](https://github.com/user-attachments/files/16651500/pprof.heap.pb.gz)
goroutine
[pprof.goroutine.pb.gz](https://github.com/user-attachments/files/16651501/pprof.goroutine.pb.gz)

All of these versions I've tested will eventually leak and system will run out of memory.

**Comment by erfantkerfan** [comment]
+1

**Comment by SasukeFreestyle** [comment]
Hello again and hope you're well :)

1.9.4 dump.
![image](https://github.com/user-attachments/assets/30da2a49-abfb-484a-8fcd-680f3b99a3e4)

heap:
![heap-crash](https://github.com/user-attachments/assets/8dcefbc7-f0ac-415b-95d0-93840bb50983)

goroutine:
![goroutine-crash](https://github.com/user-attachments/assets/21b3117a-285c-4777-aa77-1fe6da192256)

heap:
[pprof.heap.pb.gz](https://github.com/user-attachments/files/16663462/pprof.heap.pb.gz)

goroutine:
[pprof.goroutine.pb.gz](https://github.com/user-attachments/files/16663463/pprof.goroutine.pb.gz)

**Comment by SasukeFreestyle** [comment]
1.10.0-beta.2 dump

![image](https://github.com/user-attachments/assets/977fa728-7347-4180-b253-29f82be661b5)

heap:
![heap-crash](https://github.com/user-attachments/assets/41083c3a-2705-4264-8e13-3ad79a051afd)

goroutine:
![goroutine-crash](https://github.com/user-attachments/assets/2b3fcd40-a20c-4520-bc5d-602ab2c8903e)

heap:
[pprof.heap.pb.gz](https://github.com/user-attachments/files/16664541/pprof.heap.pb.gz)

goroutine:
[pprof.goroutine.011.pb.gz](https://github.com/user-attachments/files/16664528/pprof.goroutine.011.pb.gz)

**Comment by SasukeFreestyle** [comment]
Seems fixed in 1.10.0-beta.4 :)
Do you want me to dump pprof of beta.4?

If you don't need them you can close this issue.

**Comment by nekohasekai** [comment]
I don't think so, beta 4 has no changes related to this issue.

**Comment by nekohasekai** [comment]
I have no progress on your issue, we did fix a memory leak.

The new goroutine snapshots make me think that maybe you did have over 6k active UDP connections causing this occupancy, I'm not sure if there is still a leak.

If you'd like to continue the discussion via IM, you can contact me at Telegram@attachBaseContext or Discord@nekohasekai.

**Comment by SasukeFreestyle** [comment]
I honestly lack the knowledge to determine how much memory is required for lets say 6k active connections and I agree that it might not be a leak in that regard.

But after testing beta.4 a couple of days now clearly see multiple times that for example if sing-box is using 700MB, it will after a couple hours drop to around 500MB and then repeat back to 700MB and down again to 500MB.

Anyway here is 15hr uptime pprof of beta.4

I want to thank you for taking your time on this and the reason I'm doing is to help make sing-box be as efficiently as possible

I realize that no changes related to this issue is present is beta.4 but from an end-user view it is magically fixed I guess :)

![image](https://github.com/user-attachments/assets/42f246ea-401d-4c7f-bb4f-d9ecda0c0675)

heap:
![heap-fix](https://github.com/user-attachments/assets/f6fded58-9842-4e07-b7b4-628cd81f20c4)

goroutine:
![goroutine-fix](https://github.com/user-attachments/assets/be72c758-1d47-4dfb-a111-fe9996257e68)

heap:
[pprof.heap.pb.gz](https://github.com/user-attachments/files/16747443/pprof.heap.pb.gz)

goroutine:
[pprof.goroutine.pb.gz](https://github.com/user-attachments/files/16747445/pprof.goroutine.pb.gz)

**Comment by nekohasekai** [comment]
Can you try 1.9.4 again? By the way, a large number of UDP connections may be caused by protocols such as BitTorrent. You can try blocking protocol dtls and bittorrent.

**Comment by SasukeFreestyle** [comment]
I added the following lines to my configuration under rules, version 1.9.4

```json
      "rules":[
         {
            "outbound":"block",
            "protocol":[
               "bittorrent",
               "dtls"
            ]
         },
````
Memory usage is stable. Between 450MB and 750MB during peak-hours.

heap:
![heap-fix](https://github.com/user-attachments/assets/99802656-0fb1-45aa-898c-486ce91e6669)
goroutine:
![goroutine-fix](https://github.com/user-attachments/assets/6e003e29-3b12-490c-995d-2541354f0dc3)

heap:
[pprof.heap.pb.gz](https://github.com/user-attachments/files/16828884/pprof.heap.pb.gz)
goroutine:
[pprof.goroutine.pb.gz](https://github.com/user-attachments/files/16828885/pprof.goroutine.pb.gz)

I don't mind blocking DTLS/Bittorrent traffic and if this is a fix for the memory usage I'm happy with the results and thank you for the tip.

I just want to also point out that after adding these rules I checked if Bittorrent was blocked and it worked, Using qBittorrent
But that's another issue not related to this topic.
